govcd/nsxv_distributed_firewall.go

```diff
@@ -0,0 +1,486 @@
+	services, err := dfw.GetServices(false)
+	if err != nil {
+		return nil, err
+	}
+	var foundService types.Application
+	for _, app := range services {
+		if app.Name == serviceName {
+			if foundService.ObjectID != "" {
+				return nil, fmt.Errorf("more than one service found with name '%s'", serviceName)
+			}
+			foundService = app
+		}
+	}
+	if foundService.ObjectID == "" {
+		return nil, ErrorEntityNotFound
+	}
+	return &foundService, nil
+}
+
+// GetServicesByRegex returns a list of services with their names matching the given regular expression
+// It may return an empty list (without error)
+func (dfw *NsxvDistributedFirewall) GetServicesByRegex(expression string) ([]types.Application, error) {
+	services, err := dfw.GetServices(false)
+	if err != nil {
+		return nil, err
+	}
+	searchRegex := regexp.MustCompile(expression)
+	var found []types.Application
+	for _, app := range services {
+		if searchRegex.MatchString(app.Name) {
+			found = append(found, app)
+		}
+	}
+	return found, nil
+}
+
+// GetServiceGroupById retrieves a single service group, identified by its ID, for the current VDC
+// If the list of service groups was already retrieved, it uses it, otherwise fetches new ones.
+// Returns ErrorEntityNotFound when the requested service group was not found
+func (dfw *NsxvDistributedFirewall) GetServiceGroupById(serviceGroupId string) (*types.ApplicationGroup, error) {
+	serviceGroups, err := dfw.GetServiceGroups(false)
+	if err != nil {
+		return nil, err
+	}
+	for _, appGroup := range serviceGroups {
+		if appGroup.ObjectID == serviceGroupId {
+			return &appGroup, nil
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetServiceGroupByName retrieves a single service group, identified by its name, for the current VDC
+// If the list of service groups was already retrieved, it uses it, otherwise fetches new ones.
+// Returns ErrorEntityNotFound when the requested service group was not found
+func (dfw *NsxvDistributedFirewall) GetServiceGroupByName(serviceGroupName string) (*types.ApplicationGroup, error) {
+	serviceGroups, err := dfw.GetServiceGroups(false)
+	if err != nil {
+		return nil, err
+	}
+	var foundAppGroup types.ApplicationGroup
+	for _, appGroup := range serviceGroups {
+		if appGroup.Name == serviceGroupName {
+			if foundAppGroup.ObjectID != "" {
+				return nil, fmt.Errorf("more than one service group found with name %s", serviceGroupName)
+			}
+			foundAppGroup = appGroup
+		}
+	}
+	if foundAppGroup.ObjectID == "" {
+		return nil, ErrorEntityNotFound
+	}
+	return &foundAppGroup, nil
+}
+
+// GetServiceGroupsByRegex returns a list of services with their names matching the given regular expression
+// It may return an empty list (without error)
+func (dfw *NsxvDistributedFirewall) GetServiceGroupsByRegex(expression string) ([]types.ApplicationGroup, error) {
+	serviceGroups, err := dfw.GetServiceGroups(false)
+	if err != nil {
+		return nil, err
+	}
+	searchRegex := regexp.MustCompile(expression)
```

Comment:
Given that this `expression` is supplied by user - IMO, we should use `regexp.Compile` and return the error to user if is invalid

Reply:
Thanks. Fixed